### PR TITLE
[WIP] change how qualified types are handled

### PR DESCRIPTION
**Pull request by datho7561**

### CODE REVIEW

File: org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/CompilationUnit.java [hunk 1]
```diff
@@ -452,6 +454,20 @@ public IJavaElement[] codeSelect(int offset, int length, WorkingCopyOwner workin
...
+		ASTNode parentFinder = node.getParent();
+		while (parentFinder.getParent() != null && parentFinder.getParent() instanceof Type) {
+			parentFinder = parentFinder.getParent();
+		}
+		if (parentFinder.getParent() != null && parentFinder.getParent() instanceof ClassInstanceCreation) {
+			return parentFinder.getParent();
+		}
+	}
+	return node;
+}
```
Comment (mickaelistria) [inline org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/CompilationUnit.java:468]: There is a trick here: imagine one type `new SomeParentType.SomeNestedType((SomeRandomOtherType)null, (YetAnotherRandomOtherType)null)`, then with the change, all the types would then lead to their constructor definition, while in practice, we'd only want it if selecting the `SomeNestedType`. So additional checks are necessary.

**Review by mickaelistria — CHANGES_REQUESTED** [review]

**Comment by mickaelistria** [comment]
The first change, fixing Util.getSignature() is a JDT bug. Can you please report it to JDT and make it a JDT PR? Then add it to our "blockers" tracker: https://github.com/orgs/eclipse-jdtls/projects/1/views/1 ?

**Comment by mickaelistria** [comment]
The part about constructor binding was addressed in https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/92
The bit about Signature should be reported and considered upstream (and then we'll cherry-pick it)

**Comment by mickaelistria** [comment]
As I've merged an equivalent patch that includes support for Constructor special case, I've rebased the qualifier patch and submitted it in a dedicated PR to get an idea of what it fixes: https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/96

**Comment by datho7561** [comment]
> Then add it to our "blockers" tracker: https://github.com/orgs/eclipse-jdtls/projects/1/views/1 ?

@mickaelistria here is the upstream issue: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2074 .
I can't seem to add it to the board, I probably would need committership on eclipse.jdt.ls in order to do that.

**Pull request closed without merging**